### ISSUE TITLE
New Tpetra and Zoltan2 test failures in all ATDM Trilinos CUDA builds starting 12/21/2018

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#4123]
CC: @trilinos/tpetra, @trilinos/zoltan2 , @kddevin (Trilinos Data Services Product Lead), @bartlettroscoe, @fryeguy52, @tjfulle 

<Checklist>
<???: Once GitHub Issue is created, add entries for tests to TrilinosATDMStatus/*.csv files>

## Next Action Status

PR #4124 that reverts PR #4070 was merged to 'develop' on 12/21/2018.  Next: Watch for these failures to clean up on 12/22/2018 ...

## Description

As shown in [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2018-12-21&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=buildname&compare2=62&value2=Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug-pt&field3=buildname&compare3=63&value3=cuda&field4=status&compare4=62&value4=passed&field5=status&compare5=62&value5=notrun) the tests:

* TpetraCore_CrsGraph_insertGlobalIndicesFiltered_MPI_2
* TpetraCore_CrsGraph_StaticImportExport_MPI_4
* TpetraCore_CrsGraph_UnitTests0_MPI_4
* TpetraCore_CrsGraph_UnpackIntoStaticGraph_MPI_4
* TpetraCore_Issue601_MPI_4
* Zoltan2_TpetraRowGraphInput_MPI_4
* Zoltan2_XpetraCrsGraphInput_MPI_4
* Zoltan2_XpetraTraits_MPI_4

are newly failing in all of the ATDM Trilinos CUDA builds:

* Trilinos-atdm-waterman-cuda-9.2-debug
* Trilinos-atdm-waterman-cuda-9.2-opt
* Trilinos-atdm-waterman-cuda-9.2-release-debug
* Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-debug
* Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release
* Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug

Looking at the test output for the failing tests in the build `Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug` most of the failing tests shown errors like [TpetraCore_CrsGraph_insertGlobalIndicesFiltered_MPI_2](https://testing.sandia.gov/cdash/testDetails.php?test=62289874&build=4336759) showing:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/jenkins/white/workspace/Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug/SRC_AND_BUILD/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Impl.cpp:119
Traceback functionality not available

[white22:69850] *** Process received signal ***
[white22:69850] Signal: Aborted (6)
[white22:69850] Signal code:  (-6)
```
and like [TpetraCore_CrsGraph_UnpackIntoStaticGraph_MPI_4](https://testing.sandia.gov/cdash/testDetails.php?test=62290045&build=4336759) showing:

```
block: [0,0,0], thread: [0,0,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,1,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,2,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,3,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaMemcpy( dst , src , n , cudaMemcpyDefault ) error( cudaErrorAssert): device-side assert triggered /home/jenkins/white/workspace/Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug/SRC_AND_BUILD/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:89
Traceback functionality not available

[white22:75371] *** Process received signal ***
[white22:75371] Signal: Aborted (6)
[white22:75371] Signal code:  (-6)
```

Pretty much all of the errors except the for the test `TpetraCore_CrsGraph_UnpackIntoStaticGraph_MPI_4` shown above show:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered ...
```

The new commits that were pulled the day that these failures started are show, for example, [here](https://testing.sandia.gov/cdash-dev-view/viewNotes.php?buildid=4336672#!#note6).  From looking over that set of commits, it seems very likely that the commit 9ec2271 from @tjfulle merged in from PR #4070 on 12/20/2018 (yesterday) is the cause.

## Current Status on CDash

The current status of these CUDA builds can be found at:

* [ATDM Trilinos CUDA Builds for current testing day](https://testing.sandia.gov/cdash-dev-view/index.php?project=Trilinos&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=buildname&compare2=63&value2=cuda&field3=buildname&compare3=62&value3=Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug-pt&field4=site&compare4=62&value4=ride)
* [ATDM Trilinos CUDA Builds non-passing tests for current testing day](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=6&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=buildname&compare2=62&value2=Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug-pt&field3=buildname&compare3=63&value3=cuda&field4=site&compare4=62&value4=ride&field5=status&compare5=62&value5=passed&field6=status&compare6=62&value6=notrun)

## Steps to Reproduce

One should be able to reproduce this failure on the any of the machines with a CUDA build as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system 'white' or ''ride' are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ridewhite

The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh cuda-9.2-gnu-7.2.0-release-debug

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Tpetra=ON  -DTrilinos_ENABLE_Zoltan2=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -q rhel7F -n 16 ctest  ctest -j16
```

**Comment by bartlettroscoe** [comment]
@kddevin, @tjfulle, @mhoemmen 

I manually verified that reverting the commit 9ec2271 from PR #4070 makes the errors go away (see details below).

Given that it looks like there are as many as 50 test failures triggered across many CUDA builds and that fact that we are entering the XMAS break, can we just revert this commit 9ec2271 and then fix this offline in the new year?

<details>
<summary><b>DETAILED NOTES:</b> (Click to expand)</summary>

Reproducing the failure on 'white' ...

For the Trilinos 'develop' version:

```
7db87be "Merge Pull Request #4120 from william76/Trilinos/Tpetra_CrsGraph_Swap_Cleanup"
Author: trilinos-autotester <trilinos-autotester@trilinos.org>
Date:   Thu Dec 20 19:55:45 2018 -0700 (11 hours ago)
```

On 'white' I ran:

```
$ bsub -x -I -q rhel7F -n 16 \
  ./checkin-test-atdm.sh cuda-9.2-gnu-7.2.0-release-debug \
  --enable-packages=Tpetra,Zoltan2 --local-do-all
```

which returned the test results:

```
 97% tests passed, 8 tests failed out of 295
  
  Subproject Time Summary:
  Tpetra     = 1138.85 sec*proc (194 tests)
  Zoltan2    = 571.96 sec*proc (101 tests)
  
  Total Test time (real) = 216.76 sec
  
  The following tests FAILED:
  	 41 - TpetraCore_CrsGraph_UnitTests0_MPI_4 (Failed)
  	 45 - TpetraCore_Issue601_MPI_4 (Failed)
  	 46 - TpetraCore_CrsGraph_insertGlobalIndicesFiltered_MPI_2 (Failed)
  	 50 - TpetraCore_CrsGraph_UnpackIntoStaticGraph_MPI_4 (Failed)
  	 51 - TpetraCore_CrsGraph_StaticImportExport_MPI_4 (Failed)
  	210 - Zoltan2_TpetraRowGraphInput_MPI_4 (Failed)
  	211 - Zoltan2_XpetraTraits_MPI_4 (Failed)
  	213 - Zoltan2_XpetraCrsGraphInput_MPI_4 (Failed)
  Errors while running CTest
  
  Total time for cuda-9.2-gnu-7.2.0-release-debug = 21.05 min
```

Now, reverting the commit 9ec2271:

```
$ cd Trilinos/

$ git revert 9ec2271
```

and then running the build and tests again:

```
$ bsub -x -I -q rhel7F -n 16 \
  ./checkin-test-atdm.sh cuda-9.2-gnu-7.2.0-release-debug \
  --enable-packages=Tpetra,Zoltan2 --local-do-all
```

which returned:

```
  100% tests passed, 0 tests failed out of 293
  
  Subproject Time Summary:
  Tpetra     = 1172.07 sec*proc (192 tests)
  Zoltan2    = 542.94 sec*proc (101 tests)
  
  Total Test time (real) = 217.46 sec
  
  Total time for cuda-9.2-gnu-7.2.0-release-debug = 15.94 min
```

So that proves that it was the commit 9ec2271 that triggered these new CUDA failures.

</details>

**Comment by bartlettroscoe** [comment]
FYI: Newly failing/timing-out tests:

* `Ifpack2_unit_tests_MPI_4` failing in build `Trilinos-atdm-waterman-cuda-9.2-opt`
* `PanzerDofMgr_tUniqueGlobalIndexerUtilities_MPI_2` timing-out in build `Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-debug`

shown in [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2018-12-21&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=6&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=buildname&compare2=62&value2=Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug-pt&field3=buildname&compare3=63&value3=cuda&field4=site&compare4=62&value4=ride&field5=status&compare5=62&value5=passed&field6=status&compare6=62&value6=notrun) may have also been triggered by commit  9ec2271?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe, yes, let’s revert the commit.im away from a computer at the moment (on mobile), can you revert it? Otherwise, I can revert tonight.

@tjfulle, yes, I can revert.  I will put on a PR shortly.

**Comment by bartlettroscoe** [comment]
@tjfulle, I posted PR #4124.  Can you please approve that?  (You can do that on your phone browser.)

Otherwise, @mhoemmen, can you approve PR #4124?

**Comment by bartlettroscoe** [comment]
PR #4124 that reverts PR #4070 was merged to 'develop' on 12/21/2018.  Next: Watch for these failures to clean up on 12/22/2018 ...

Putting "in review" ...

**Comment by tjfulle** [comment]
PR #4147 fixes the Cuda error in #4070 in my testing on `ride` using the ATDM environment.

**Comment by bartlettroscoe** [comment]
Reverging PR #2464 merged to 'develop' on 12/21/2019 resolved these failures.  Closing as complete.
